[JSC][armv7] Enable concurrent JIT

jjgriego opened the pull request:
#### 0e5f4888c8bb1f7d0a4407d044089f664bc857df
<pre>
[JSC][armv7] Enable concurrent JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=239821">https://bugs.webkit.org/show_bug.cgi?id=239821</a>

Reviewed by NOBODY (OOPS!).

Here we are again--the goal is to enable concurrent JIT support on ARMv7--the
problem, of course, is we don&apos;t have a way to update a JSValue atomically, with
the current encoding.

** Approach

Previous passes at this have tried to cope with the possibility of reading a
&quot;spliced JSValue&quot;--i.e., a value observed with a CellTag and a not-Cell
payload; this seems to work but isn&apos;t great and we don&apos;t have a way to get
reasonable answers from e.g. value profiles when this happens.

Instead, we develop the following protocol for updating a JSValue:

- Write a newly-reserved tag value, `InvalidTag` to the tag part of the value,
- Write the new payload word
- Write the new tag word

Now, when _reading_ a JSValue from threads other than the mutator, we can use
doubleword atomics and retry until the tag value is not `InvalidTag`.

Unfortunately, although in practice I&apos;ve been unable to find any hardware where
store reordering seems to be observable, the architecture requires us to insert
store-store fences (`dmb ishst`) between these writes; this isn&apos;t free, as
discussed below.

** Implementation

We add a few new primitives for manipulating JSValues in memory that are used in
some key places where JSValues are accessed concurrently: ValueProfiles and any
JSValue behind a WriteBarrier (which addresses most things in the heap.)

As discussed above, for writes, we perform the 3 fenced stores, in order; reads
are retried until the tag is not InvalidTag; and a
`clearEncodedJSValueConcurrent` is provided since this operation does not
require fences (only the tag is modified)

** Performance

The addition of fences adds a nontrivial performance cost, especially for code
that spends a lot of time in the interpreter; however, this is still an
improvement for JetStream2 first-run times on some benchmarks, overall, the
performance is about the same for JS2.

I expect a regression to code size because of the extra stores--though I
believe there&apos;s opportunity to optimize some of these away in the JITs based on
speculation--if we know the JSValue is already a cell or non-cell, the
InvalidTag store is unnecssary and can be omitted.

* Source/JavaScriptCore/bytecode/ValueProfile.h:
(JSC::ValueProfileBase::clearBuckets):
(JSC::ValueProfileBase::classInfo const):
(JSC::ValueProfileBase::numberOfSamples const):
(JSC::ValueProfileBase::isLive const):
(JSC::ValueProfileBase::computeUpdatedPrediction):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::storeAndFence32):
(JSC::AssemblyHelpers::storeCell):
(JSC::AssemblyHelpers::storeValue):
(JSC::AssemblyHelpers::storeTrustedValue):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/offlineasm/arm.rb:
* Source/JavaScriptCore/offlineasm/instructions.rb:
* Source/JavaScriptCore/runtime/JSCJSValue.h:
(JSC::JSValue::decodeConcurrent):
(JSC::updateEncodedJSValueConcurrent):
(JSC::clearEncodedJSValueConcurrent):
* Source/JavaScriptCore/runtime/WriteBarrier.h:
* Source/JavaScriptCore/runtime/WriteBarrierInlines.h:
(JSC::RawValueTraits&lt;Unknown&gt;&gt;::set):
* Source/WTF/wtf/PlatformEnable.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e5f4888c8bb1f7d0a4407d044089f664bc857df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26400 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5013 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27660 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28552 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24163 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6816 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2441 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24154 "27 flakes 144 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26660 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3870 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22703 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3426 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3459 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23680 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29097 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23017 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24084 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24081 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29739 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25614 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3473 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1698 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27633 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4899 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33061 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3954 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/7190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3798 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
```diff
@@ -263,7 +275,28 @@ inline JSValue JSValue::decode(EncodedJSValue encodedJSValue)
+inline JSValue JSValue::decodeConcurrent(const volatile EncodedJSValueStorage *encodedJSValue)
+{
+    for (;;) {
+        auto v = JSValue::decode(reinterpret_cast<const volatile std::atomic<EncodedJSValue>*>(encodedJSValue)->load());
```
What is the actual sequence of instructions generated for this code? Can you paste the asm here?

jjgriego replied:
Sure:

```
Dump of assembler code for function _ZN3JSC7JSValue16decodeConcurrentEPVKNS_21EncodedJSValueStorageE:
   0x00415644 <+0>:     ldrexd  r2, r3, [r1]
   0x00415648 <+4>:     dmb     ish
   0x0041564c <+8>:     cmn.w   r3, #9
   0x00415650 <+12>:    beq.n   0x415644 <_ZN3JSC7JSValue16decodeConcurrentEPVKNS_21EncodedJSValueStorageE>
   0x00415652 <+14>:    strd    r2, r3, [r0]
   0x00415656 <+18>:    bx      lr
```

Constellation commented on Source/JavaScriptCore/runtime/JSCJSValue.h:
```diff
@@ -86,8 +86,87 @@ enum PreferredPrimitiveType : uint8_t { NoPreference, PreferNumber, PreferString
+ * of the whole JSValue and retry if they observe the reserved state.
+ */
+#if USE(JSVALUE32_64) && ENABLE(JIT)
+struct alignas(8) EncodedJSValueStorage {
```
Why not attaching `alignas(8)` to EncodedJSValue in 32bit environment? This makes code much cleaner.

jjgriego replied:
Yeah, it's unfortunate how much extra noise there is because this type isn't POD anymore.

Alignment alone isn't enough, though--I think we need runtime code that is touching JSValues in the heap to do the extra write of InvalidTag (and to perform the writes on `volatile` int32_ts); As far as I could tell, using this type as storage in WriteBarrier and friends achieves this nicely; I suppose an alternative approach would be to move this code to those places?

Constellation replied:
> I suppose an alternative approach would be to move this code to those places?

I think that is better. We should minimize the use of`USE(JSVALUE32_64)` ifdefs as much as possible.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.h:
```diff
@@ -144,38 +144,68 @@ class AssemblyHelpers : public MacroAssembler {
-#endif
+        store32(TrustedImm32(JSValue::InvalidTag), address.withOffset(TagOffset));
+        store32(cell, address);
+        store32(TrustedImm32(JSValue::CellTag), address.withOffset(TagOffset));
```
So, what guarantees the ordering of these three stores? ARM v7 deploys weak memory ordering and stores can be reordered freely. This means that the other core can see the memory, only cell is written. (InvalidTag and CellTag stores are reordered and deferred after the cell store)

jjgriego replied:
The ISA does require the stores to occur in program order as far as the core performing them is concerned; we're relying on the hardware's cache coherency to do the rest; as long as the stores reach the L1 cache in program order (and, IIUC they must: there's only one L1 cache line involved, reordering them in general might violate the ISA requirement), ... then we'll observe them that way on other cores. 

That is to say, it's hardware-specific and, you're right, not guaranteed by the ISA.

Admittedly, I just double-checked and some older ARMv7 chips (e.g. Cortex-A8) probably _don't_ provide the cache coherency guarantees we're using here; but we only care about Cortex-A15 and some armv8 chips (A53, A55) which do... I think we'll probably need a build flag to control whether or not this code is used.

We've tested the behavior of the store ordering on hardware: at least on a Cortex-A72 and also on our buildbox (which has a Neoverse N-1 ....) it behaves as expected. I have a board with a Cortex-A8 I can test on; but, based on the reference from ARM I'm not sure if it will work--though, because we don't really have multiple cores writing, we might luck out--it's hard for me to say without testing... I'll update here.

Constellation replied:
I think it should be checked that the hardware model of CPU and disable concurrent JIT when it is ARMv7 and unsupported cores in Options.cpp. Also, can you add a comment to Options.cpp's disabling / enabling code, which comment points to some references of this behavior (like, the URL describing this behavior for a certain cores, like hardware spec text).

eugeneia commented on Source/JavaScriptCore/runtime/CachedTypes.cpp:
```diff
@@ -1383,7 +1383,7 @@ class CachedJSValue : public VariableLengthObject<WriteBarrier<Unknown>> {
         switch (m_type) {
         case EncodedType::JSValue:
-            v = JSValue::decode(*this->buffer<EncodedJSValue>());
+            v = JSValue::decode(*this->buffer<EncodedJSValueStorage>()); // XXX - EncodedJSValue or decodeConcurrent?
```
Eh one of my surveying comments snuck in here.

jjgriego committed:
[JSC][armv7] Enable concurrent JIT
https://bugs.webkit.org/show_bug.cgi?id=239821

Reviewed by NOBODY (OOPS!).

Here we are again--the goal is to enable concurrent JIT support on ARMv7--the
problem, of course, is we don't have a way to update a JSValue atomically, with
the current encoding.

** Approach

Previous passes at this have tried to cope with the possibility of reading a
"spliced JSValue"--i.e., a value observed with a CellTag and a not-Cell
payload; this seems to work but isn't great and we don't have a way to get
reasonable answers from e.g. value profiles when this happens.

Instead, we develop the following protocol for updating a JSValue:

- Write a newly-reserved tag value, `InvalidTag` to the tag part of the value,
- Write the new payload word
- Write the new tag word

Now, when _reading_ a JSValue from threads other than the mutator, we can use
doubleword atomics and retry until the tag value is not `InvalidTag`.

Unfortunately, although in practice I've been unable to find any hardware where
store reordering seems to be observable, the architecture requires us to insert
store-store fences (`dmb ishst`) between these writes; this isn't free, as
discussed below.

** Implementation

We add a few new primitives for manipulating JSValues in memory that are used in
some key places where JSValues are accessed concurrently: ValueProfiles and any
JSValue behind a WriteBarrier (which addresses most things in the heap.)

As discussed above, for writes, we perform the 3 fenced stores, in order; reads
are retried until the tag is not InvalidTag; and a
`clearEncodedJSValueConcurrent` is provided since this operation does not
require fences (only the tag is modified)

** Performance

The addition of fences adds a nontrivial performance cost, especially for code
that spends a lot of time in the interpreter; however, this is still an
improvement for JetStream2 first-run times on some benchmarks, overall, the
performance is about the same for JS2.

I expect a regression to code size because of the extra stores--though I
believe there's opportunity to optimize some of these away in the JITs based on
speculation--if we know the JSValue is already a cell or non-cell, the
InvalidTag store is unnecssary and can be omitted.

* Source/JavaScriptCore/bytecode/ValueProfile.h:
(JSC::ValueProfileBase::clearBuckets):
(JSC::ValueProfileBase::classInfo const):
(JSC::ValueProfileBase::numberOfSamples const):
(JSC::ValueProfileBase::isLive const):
(JSC::ValueProfileBase::computeUpdatedPrediction):
* Source/JavaScriptCore/jit/AssemblyHelpers.h:
(JSC::AssemblyHelpers::storeAndFence32):
(JSC::AssemblyHelpers::storeCell):
(JSC::AssemblyHelpers::storeValue):
(JSC::AssemblyHelpers::storeTrustedValue):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/offlineasm/arm.rb:
* Source/JavaScriptCore/offlineasm/instructions.rb:
* Source/JavaScriptCore/runtime/JSCJSValue.h:
(JSC::JSValue::decodeConcurrent):
(JSC::updateEncodedJSValueConcurrent):
(JSC::clearEncodedJSValueConcurrent):
* Source/JavaScriptCore/runtime/WriteBarrier.h:
* Source/JavaScriptCore/runtime/WriteBarrierInlines.h:
(JSC::RawValueTraits<Unknown>>::set):
* Source/WTF/wtf/PlatformEnable.h:
